s4/s4.py

```diff
@@ -335,185 +411,228 @@ def setup(self):
+# > lot of clever math that applies under a structured parameterization of the
+# > model. Specifically:
 
+# The S4 techniques apply to any matrix $\boldsymbol{A}$ that can be decomposed as *Normal Plus Low-Rank (NPLR)*.
 # $$
-# \begin{aligned}
-# \mathbf{K} &= (\mathbf{\bar{C}} \mathbf{\bar{B}}, \mathbf{\bar{C}} \mathbf{\bar{A}}^1 \mathbf{\bar{B}}, \ldots, \mathbf{\bar{C}} \mathbf{\bar{A}}^{L-1} \mathbf{\bar{B}}) \\
-# \end{aligned}
+#   \boldsymbol{A} = \boldsymbol{V} \boldsymbol{\Lambda} \boldsymbol{V}^* - \boldsymbol{p} \boldsymbol{q}^\top = \boldsymbol{V} \left( \boldsymbol{\Lambda} - \boldsymbol{V}^* \boldsymbol{p} (\boldsymbol{V}^*\boldsymbol{q})^* \right) \boldsymbol{V}^*
 # $$
+# for unitary $\boldsymbol{V} \in \mathbb{C}^{N \times N}$, diagonal $\boldsymbol{\Lambda}$, and low-rank factorization $\boldsymbol{p}, \boldsymbol{q} \in \mathbb{R}^{N \times r}$.  An NPLR SSM is therefore unitarily equivalent to some Diagonal Plus Low Rank (DPLR) $(\boldsymbol{\Lambda} - \boldsymbol{p}\boldsymbol{q}^*, \boldsymbol{B}, \boldsymbol{C})$ for some diagonal $\boldsymbol{\Lambda}$ and vectors $\boldsymbol{p}, \boldsymbol{q}, \boldsymbol{B}, \boldsymbol{C} \in \mathbb{C}^{N \times 1}$.
 
 
-# instead as a polynomial $\mathbf{\hat{K}}$ where each coefficient represents one element of this sequence.
+# Under this DPLR assumption, we can overcome this speed bottleneck by
+# simultaneously applying three new techniques.
+#
+#  1.  Instead of computing $\boldsymbol{\overline{K}}$ directly,
+#     we compute its spectrum by evaluating its **truncated generating function**  at the roots of unity.
+#     $\boldsymbol{\overline{K}}$ can then be found by applying an inverse FFT.  This generating function is closely related to the matrix resolvent, and now involves a matrix *inverse* instead of *power*.
+#  2. We show that the diagonal matrix case is equivalent to the computation of a **Cauchy kernel** $\frac{1}{\omega_j - \zeta_k}$.
+#  3. We show the low-rank term can now be corrected by applying the **Woodbury identity** which reduces $(\boldsymbol{\Lambda} + \boldsymbol{p}\boldsymbol{q}^*)^{-1}$ in terms of $\boldsymbol{\Lambda}^{-1}$, truly reducing to the diagonal case.
+#
 
-# $$ \begin{aligned}
-# \mathbf{\hat{K}}(z) = \mathbf{\bar{C}} \mathbf{\bar{B}} + \mathbf{\bar{C}} \mathbf{\bar{A}}^1 \mathbf{\bar{B}} z^1 + \ldots + \mathbf{\bar{C}} \mathbf{\bar{A}}^{L-1} \mathbf{\bar{B}} z^{l-1} \\
-# \end{aligned}
-# $$
+# Finally we note the all HiPPO matrices have this NPLR representation. We can therefore find extract a unitarily equivalent DPLR parameterization.
+
+
+# ## Step 1. SSM Generating Functions
+
+# To address the problem of computing powers of $\boldsymbol{\overline{A}}$, we introduce another technique.
+# Instead of computing the SSM convolution filter $\boldsymbol{\overline{K}}$ directly,
+# we introduce a [generating function]() on its coefficients and compute evaluations of it.
```

Comment:
I really liked this link for helping get intuition: https://math.stackexchange.com/questions/3213142/root-of-unity-filter